src/atc/sql/SqlServer.py

```diff
@@ -100,6 +104,48 @@ def write_table_by_name(
             "password", self.password
         ).save()
 
+    def upsert_to_table_by_name(
+        self,
+        df_source: DataFrame,
+        table_name: str,
+        join_cols: List[str],
+        big_data_set: bool = True,
+        batch_size: int = 10 * 1024,
+        partition_count: int = 60,
+    ):
+
+        staging_table_name = f"{table_name}_{uuid.uuid4().hex}"
+
+        # Create staging table for merge from source table
+        self.execute_sql(
+            f"SELECT * INTO {staging_table_name} FROM {table_name} WHERE 1 = 0;"
+        )
+
+        self.write_table_by_name(
+            df_source=df_source,
+            table_name=staging_table_name,
+            append=False,
+            big_data_set=big_data_set,
+            batch_size=batch_size,
+            partition_count=partition_count,
+        )
+
+        mergeQuery = GetMergeStatement(
+            merge_statement_type="sql",
+            target_table_name=table_name,
+            source_table_name=staging_table_name,
+            join_cols=join_cols,
+            insert_cols=df_source.columns,
+            update_cols=df_source.columns,
+        )
+
+        # Merge query is combined with drop staging table
+        # to ensure the drop of the staging table in case
+        # the problem stops in the midle of the upsert
+        mergeQuery += f" DROP TABLE {staging_table_name};"
```

Comment:
should the mergestatement + drop table be in a transation statement? https://docs.microsoft.com/en-us/sql/t-sql/language-elements/begin-transaction-transact-sql?view=sql-server-ver16

Reply:
I want to be sure that the table is dropped. I have seen examples where the job is stoppd in the middle of the merge(ex when stopping a stream). In this case the case the staging tables would pile up, as this happens.